Support MultiNodeMultiWriter for block volumes (kubevirt)

tiagolobocastro opened the pull request:
test(csi/rwx): extend bdd with rwx block
    
    Attempts to republish volume to another node with rwx block
    todo: add negative tests
    
---

    feat(csi/controller): add rwx block enablement via volume context
    
    A new SC parameter "rwxBlock: true", may be used to enable rwx block.
    
---

    feat(kubevirt/live-migrate): support rwx for block volumes
    
    Allows multi-node writer access modes but only if the access type
    is for Block.
    This should only be enabled for applications which can handle
    having the same block device connected by multiple initiator nodes!
    
    todo: add kube-virt flag to the storage class parameters which will
    enable this feature (or global CLI flag)
    
    Allows multiple frontend nodes for the same volume target.
    Each node's nqn is added to the allowed hosts lists in the pstor and
    in the volume's nvmf target.
    
    When unpublishing, if all frontend nodes would be removed, then the
    nexus is also destroyed.
    Note that eventually when we switch to offline rebuilding, this will
    likely have to change. Nevertheless, this is equivalent to the current
    behaviour where the unpublish destroys the nexus as well.
    
---

    test(bdd): expect no error on unpublish from different node

tiagolobocastro committed:
feat(kubevirt/live-migrate): support rwx for block volumes

Allows multi-node writer access modes but only if the access type
is for Block.
This should only be enabled for applications which can handle
having the same block device connected by multiple initiator nodes!

todo: add kube-virt flag to the storage class parameters which will
enable this feature (or global CLI flag)

Allows multiple frontend nodes for the same volume target.
Each node's nqn is added to the allowed hosts lists in the pstor and
in the volume's nvmf target.

When unpublishing, if all frontend nodes would be removed, then the
nexus is also destroyed.
Note that eventually when we switch to offline rebuilding, this will
likely have to change. Nevertheless, this is equivalent to the current
behaviour where the unpublish destroys the nexus as well.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
test(bdd): expect no error on unpublish from different node

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(csi/controller): add rwx block enablement via volume context

A new SC parameter "rwxBlock: true", may be used to enable rwx block.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
test(csi/rwx): extend bdd with rwx block

Attempts to republish volume to another node with rwx block
todo: add negative tests

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

Abhinandan-Purkait commented on control-plane/agents/src/bin/core/nexus/specs.rs:
```diff
@@ -51,11 +51,11 @@ impl SpecOperationsHelper for NexusSpec {
+            //     kind: ResourceKind::Nexus,
+            //     id: self.uuid_str(),
+            //     share: state.share.to_string(),
+            // }),
```
Remove this?

tiagolobocastro replied:
I rewrote it to take into account the allowed hosts

Abhinandan-Purkait commented on control-plane/agents/src/bin/core/volume/operations.rs:
```diff
@@ -321,6 +325,48 @@ impl ResourcePublishing for OperationGuardArc<VolumeSpec> {
+                    .await;
+                return Ok(volume);
+            } else {
+                // error out?
```
Don't need to?

tiagolobocastro replied:
This is in case we don't need an update, wasn't sure if we should error out here, wdyt?

Abhinandan-Purkait replied:
I think we should not? Just make add a log and move on?

tiagolobocastro replied:
I've removed it, since we weren't checking it prior to  these changes anyway

Abhinandan-Purkait commented on control-plane/csi-driver/src/bin/controller/controller.rs:
```diff
@@ -59,22 +59,55 @@ impl CsiControllerSvc {
+                    Ok(access_mode)
+                } else {
+                    Err(Status::invalid_argument(
+                        "MultiNodeMultiWriter but RWX Block is disabled".to_string(),
```
```suggestion
                        "MultiNodeMultiWriter for RWX Block is disabled".to_string(),
```
?

tiagolobocastro replied:
hmm, but this is to say "RWX Block" is disabled? maybe

```suggestion
                        "MultiNodeMultiWriter requested but RWX Block is disabled".to_string(),
```

Abhinandan-Purkait replied:
Sounds good.

Abhinandan-Purkait commented on control-plane/csi-driver/src/bin/controller/controller.rs:
```diff
@@ -661,18 +718,16 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
             Ok(volume) => volume,
             Err(ApiClientError::ResourceNotExists { .. }) => {
-                debug!("Volume {} does not exist, not unpublishing", args.volume_id);
+                // todo: shouldn't this be a grpc NOT_FOUND?
+                debug!("Volume does not exist, not unpublishing");
```
You mean return an error with 404?

tiagolobocastro replied:
Well this is the CSI api here, so there's no 404, only gRPC status code of `NOT_FOUND`.

Abhinandan-Purkait replied:
Yes, I meant that only.

Abhinandan-Purkait commented on control-plane/csi-driver/src/bin/controller/controller.rs:
```diff
@@ -708,23 +763,22 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
         let _volume = RestApiClient::get_client()
             .get_volume(&volume_uuid)
             .await
+            // todo: this seems wrong...
```
Indeed, should have been Unavailable rather than unimplemented, or maybe aborted with xyz reason.

tiagolobocastro replied:
Yes something like that indeed... I think I'll leave as is for now, maybe can fix this on another PR

Abhinandan-Purkait commented on control-plane/csi-driver/src/bin/controller/controller.rs:
```diff
@@ -708,23 +763,22 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
             })
             .collect();
 
+        // todo: this seems wrong? See https://github.com/container-storage-interface/spec/blob/e981e2a057ca10f4a7f81289c97a4e829fd69152/spec.md?plain=1#L1502
```
Should have been none, rather than empty defaults?

tiagolobocastro replied:
As per the docs: `This RPC call SHALL return confirmed only if all the volume capabilities specified in the request are supported (see caveat below).`
But here we return ok as long as one capability is ok, which seems to violate the above?

Abhinandan-Purkait replied:
Yes I think so.

Abhinandan-Purkait commented on control-plane/csi-driver/src/bin/controller/controller.rs:
```diff
@@ -771,6 +825,7 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
                 let volume = rpc::csi::Volume {
                     volume_id: v.spec.uuid.to_string(),
                     capacity_bytes: v.spec.size as i64,
+                    // todo: these 2 seem wrong?
```
Probably because we don't store the context? Or maybe we could have just generated it, I believe it's only the device uri that goes here?

Ahh we can definitely update the content source now that we have implemented the snaps.

tiagolobocastro replied:
For the context yes, we'd have to start storing it, and no it's not the uri, that's the publish context.
Probably we'd need clarification here...

For content source, yes we should be able to set it correctly here

Abhinandan-Purkait replied:
Yeah maybe you can just add the content source. I am always a bit skeptical about changing any error handling in csi, changes the behaviour :-;

tiagolobocastro replied:
I think we should not not implement this method

Abhinandan-Purkait commented on control-plane/csi-driver/src/bin/controller/controller.rs:
```diff
@@ -799,8 +854,10 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
+        let context = CreateParams::try_from(&args.parameters)?;
+        check_volume_capabilities(&args.volume_capabilities, context.rwx_block())?;
 
+        // todo: this seems to be conflating app node vs storage node?
```
I am a bit confused here, what exactly are we trying to do here? IIRC if we set allowedTopologies on the storage class this get's populated, and it's expected to be a subset of the csi-nodes only.

So if someone does that and that csi-node happens to be a non mayastor node we return an error saying node not found?

This probably needs to be re-evaluated, the accessible topology has been mixed with data placement topology here.

tiagolobocastro replied:
I don't know what we are trying to do here...

Abhinandan-Purkait replied:
Maybe leave it for now?

tiagolobocastro replied:
I'll leave for another PR

Abhinandan-Purkait commented on control-plane/csi-driver/src/context.rs:
```diff
@@ -511,6 +518,10 @@ pub struct CreateParams {
 }
 impl CreateParams {
+    /// Get the `Parameters::RwxBlock` value.
+    pub fn rwx_block(&self) -> &Option<bool> {
```
Just wondering do we need an explicit knob as well? What happens if someone says ReadWriteMany and sets rwxBlock to false?

tiagolobocastro replied:
Then we'll fail the validation check on `check_volume_capability`.
This is needed to ensure users enable this explicitly as this should only be used with applications which handle this properly.

Abhinandan-Purkait replied:
Makes sense.

Abhinandan-Purkait reviewed (approved):
LGTM

tiagolobocastro commented:
bors merge

tiagolobocastro committed:
ci: retry apt-update

It was seen on CI that sometimes apt-update fails, so as a quick wa,
let's just retry once after 5s and see how it goes.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>
